### ISSUE TITLE
How to integrate Diesel

### ISSUE THREAD

**Issue by sporto** [graphql-rust/juniper_rocket#4]
Hi, I have been trying to integrate Diesel but I'm not clear on how to integrate the db connection with the graphql context.

I have my code here https://github.com/sporto/rocket-graphql/blob/diesel/src/main.rs

I can provide the db connection pool `conn: db::Conn` to the request handler:

```
#[post("/graphql", data = "<request>")]
fn post_graphql_handler(context: State<GraphQLContext>,
                        conn: db::Conn,
                        request: juniper_rocket::GraphQLRequest,
                        schema: State<Schema>)
                        -> juniper_rocket::GraphQLResponse {
    request.execute(&schema, &context)
}
```

But I cannot figure out how to add the r2d2 connection pool to the GraphQL Context. I just get `... cannot be shared between threads safely` if I try to add the connection inside context e.g.

```
pub struct GraphQLContext {
    pub connection: db::Conn,
}
```

Some help or an example on this would be greatly appreciated. Thanks.

**Comment by theduke** [comment]
Hey there. A single connection is not `Sync`, so it can not be shared.

You can put the `Pool` in the context and get a single connection out of the pool inside the handler (this is what I usually do).

**Comment by theduke** [comment]
If you want to have just a single connection in the context, you'll need to wrap in in a Mutex.

**Comment by sporto** [comment]
Thanks, that certainly helped. I got the db pool into the context. 

```
let dbPool = db::init_pool();

 let context = GraphQLContext {
     dbPool: dbPool,
 };
```
    
But now I don't know how to pass the schema to the context. https://github.com/sporto/rocket-graphql/blob/diesel/src/main.rs#L101

Do you have an example on how to use `Schema::new` here and pass this?

---

Also another question, my graphql handler is very noisy as it needs to deal with the pool:

```
graphql_object!(GraphQLContext: GraphQLContext as "Query" |&self| {
    description: "The root query object of the schema"

    field users(&executor) -> FieldResult<Vec<User>> {
        match executor.context().dbPool.get() {
            Ok(conn) =>
                Ok(User::all(&conn)),
            Err(_) =>
                Err(FieldError::new("DB not available", Value::null())),
        }
    }
});
```

Any ideas on how to make this more succinct?

**Comment by theduke** [comment]
@sporto we just released `0.9.0` and there is also a new book, which contains more guidance.

You can check out the quickstart guide.
Please note though that the book is still very much unfinished, but the quickstart guide is new and the code compiles. ;)

https://graphql-rust.github.io/quickstart.html

If you are still having issues, please open an issue on the main juniper repo.
